collectors/CMPCollector.js

```diff
@@ -258,6 +308,41 @@ class CMPCollector extends BaseCollector {
         }
     }
 
+    /**
+     * @returns {Promise<string[]>}
+     */
+    async scanPatterns() {
+        /**
+         * @type {string[]}
+         */
+        const found = [];
+        const pages = await this.context.pages();
+        if (pages.length > 0) {
+            const page = pages[0];
+            /**
+             * @type {Promise<string>[]}
+             */
+            const promises = [];
+            page.frames().forEach(frame => {
+                // eslint-disable-next-line no-undef
+                promises.push(frame.evaluate(() => document.documentElement.innerText).catch(reason => {
+                    this.log(`error retrieving text: ${reason}`);
+                    // ignore exceptions
+                    return '';
+                }));
+            });
+            const texts = await Promise.all(promises);
+            const allTexts = texts.join('\n');
+            for (const p of DETECT_PATTERNS) {
+                if (p.test(allTexts)) {
+                    found.push(p.toString());
```

Comment:
This is a great idea! I think it's useful to keep the patterns too, to collect [stats](https://grafana.duckduckgo.com/goto/Y-ANCm74k?orgId=1) per pattern. Shall I just add a new column?

Reply:
I figured that patterns can be reverse engineered based on the matched text, but it would make it harder to calculate those stats, so if we need stats let's keep both patterns and matched text 👍